File: backend/app/schemas/payments.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class PaymentCreate(BaseModel):
    receipt_date: date
    customer_party_id: UUID
    organization_id: UUID
    bank_account_id: UUID | None = None
    payment_mode: str = Field(pattern="^(BANK_TRANSFER|NEFT|RTGS|IMPS|UPI|CHEQUE|CASH|OTHER)$")
    transaction_reference: str | None = Field(default=None, max_length=255)
    transaction_date: date | None = None
    amount_received: Decimal = Field(gt=0, max_digits=18, decimal_places=2)
    currency: str = Field(default="INR", pattern="^[A-Z]{3}$")
    notes: str | None = Field(default=None, max_length=5000)

    @model_validator(mode="after")
    def validate_reference(self):
        if self.payment_mode in {"BANK_TRANSFER", "NEFT", "RTGS", "IMPS", "UPI", "CHEQUE"}:
            if not self.transaction_reference:
                raise ValueError("Transaction/reference number is required for this payment mode.")
        return self


class PaymentUpdate(BaseModel):
    receipt_date: date | None = None
    bank_account_id: UUID | None = None
    payment_mode: str | None = Field(
        default=None, pattern="^(BANK_TRANSFER|NEFT|RTGS|IMPS|UPI|CHEQUE|CASH|OTHER)$"
    )
    transaction_reference: str | None = Field(default=None, max_length=255)
    transaction_date: date | None = None
    amount_received: Decimal | None = Field(default=None, gt=0, max_digits=18, decimal_places=2)
    notes: str | None = Field(default=None, max_length=5000)


class AllocationCreate(BaseModel):
    tax_invoice_id: UUID
    allocated_amount: Decimal = Field(gt=0, max_digits=18, decimal_places=2)
    allocation_date: date
    remarks: str | None = Field(default=None, max_length=1000)


class PaymentAction(BaseModel):
    action: str = Field(pattern="^(CONFIRM|CANCEL|REVERSE)$")
    reason: str = Field(min_length=1, max_length=1000)


class AllocationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    payment_id: UUID
    tax_invoice_id: UUID
    allocated_amount: Decimal
    allocation_date: date
    actor_user_id: UUID
    remarks: str | None
    invoice_number_snapshot: str
    project_snapshot: str | None
    loa_snapshot: str | None
    created_at: datetime


class PaymentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    receipt_number: str
    receipt_date: date
    customer_party_id: UUID
    organization_id: UUID
    bank_account_id: UUID | None
    payment_mode: str
    transaction_reference: str | None
    transaction_date: date | None
    amount_received: Decimal
    currency: str
    notes: str | None
    status: str
    customer_snapshot: dict
    organization_snapshot: dict
    bank_snapshot: dict | None
    created_by_user_id: UUID
    confirmed_by_user_id: UUID | None
    confirmed_at: datetime | None
    reversed_by_user_id: UUID | None
    reversed_at: datetime | None
    reversal_reason: str | None
    created_at: datetime
    updated_at: datetime
    allocations: list[AllocationResponse]
    allocated_amount: Decimal
    unallocated_amount: Decimal


class ReceivablePosition(BaseModel):
    tax_invoice_id: UUID
    invoice_number: str
    invoice_date: date
    customer_party_id: UUID
    customer_name: str
    project_id: UUID
    project_name: str | None
    loa_id: UUID | None
    loa_number: str | None
    railway_division_id: UUID | None
    due_date: date | None
    invoice_total: Decimal
    received_amount: Decimal
    outstanding_amount: Decimal
    payment_status: str
    days_overdue: int


class EligibleInvoice(BaseModel):
    tax_invoice_id: UUID
    invoice_number: str
    invoice_date: date
    due_date: date | None
    invoice_total: Decimal
    received_amount: Decimal
    outstanding_amount: Decimal
    project_name: str | None
    loa_number: str | None


class InvoicePaymentHistory(BaseModel):
    payment_id: UUID
    receipt_number: str
    receipt_date: date
    payment_status: str
    allocated_amount: Decimal
    allocation_date: date
    remarks: str | None

File: frontend/src/api/payments.ts
import { apiRequest } from "./client";

export type Allocation = { id: string; tax_invoice_id: string; allocated_amount: string; invoice_number_snapshot: string };
export type Payment = { id: string; receipt_number: string; receipt_date: string; customer_party_id: string; organization_id: string; payment_mode: string; transaction_reference: string | null; amount_received: string; status: string; customer_snapshot: Record<string, string>; allocations: Allocation[]; allocated_amount: string; unallocated_amount: string };
export type Receivable = { tax_invoice_id: string; invoice_number: string; customer_name: string; project_name: string | null; loa_number: string | null; due_date: string | null; invoice_total: string; received_amount: string; outstanding_amount: string; payment_status: string; days_overdue: number };
export type EligibleInvoice = { tax_invoice_id: string; invoice_number: string; outstanding_amount: string; project_name: string | null };

export const listPayments = (token: string) => apiRequest<Payment[]>("/payments", { token });
export const listReceivables = (token: string, query = "") => apiRequest<Receivable[]>(`/receivables${query ? `?${query}` : ""}`, { token });
export const eligibleInvoices = (token: string, id: string) => apiRequest<EligibleInvoice[]>(`/payments/${id}/eligible-invoices`, { token });
export const createPayment = (token: string, body: unknown) => apiRequest<Payment>("/payments", { method: "POST", token, body: JSON.stringify(body) });
export const allocatePayment = (token: string, id: string, body: unknown) => apiRequest<Payment>(`/payments/${id}/allocations`, { method: "POST", token, body: JSON.stringify(body) });
export const paymentAction = (token: string, id: string, action: string, reason: string) => apiRequest<Payment>(`/payments/${id}/actions`, { method: "POST", token, body: JSON.stringify({ action, reason }) });
